feat: PHP functions generator

M4tteoP opened the pull request:
Continues https://github.com/coreruleset/coreruleset/pull/3228.

Opening this draft PR to show the current status of the work and get some feedback based on the diffs on `933161.ra`, `php-function-names-933150.data`, and `php-function-names-933151.data` files.

### The Script
- The script takes about **5** hours to retrieve all the frequencies. I don't think we want to merge the `frequencylist.txt`, still committed it if anyone wants to play with it (Run command example `./php-dictionary-creator.sh --frequencylist frequencylist.txt --verbose`).
- Updating `933161.ra` requires running the toolchain. The script currently only prints a reminder about it. If we feel that it is in the scope of this script, this call could be also automated.

### Rule **933160**
- The script currently does basically nothing about `933160`. 🚧 It requires some manual work to update it.
- I was taking into account adding to `933160` the PHP functions interpreted as English words (`933161`) into `933160` but it does not seem right to me considering that `933161` is the stricter sibling of the two. We should maybe do it the other way around (adding to `933161` also the manual list we have in `933160`)

### Rule **933161**
- As already reported by [Christian](https://github.com/coreruleset/coreruleset/pull/3228#issuecomment-1604929859), currently this list is way smaller than the current one. But it will be increased by:
   - 🚧 Running the script with the [`spell.sh extended`](https://github.com/coreruleset/coreruleset/pull/3228#issuecomment-1607026192)
   - If `933160` list has also to be included in `933161`

### Rules **933150** and **933151**
- `933150` would become way bigger with lots of `933151`words moved into `933150`. The decision about putting a word under one or the other is based on the `FREQUENCY_LIMIT`, set by default at `100`. 🚧 It might have to be fine-tuned.
For example, setting it at `150` moves out of `933150`:
```
-collator_compare
-curl_pause
-curl_share_init
-curl_share_setopt
-curl_share_strerror
-curl_unescape
-gmp_binomial
-imageaffine
-imageavif
-imagegetclip
-inflate_get_read_len
-ldap_exop_refresh
-ldap_parse_exop
-locale_get_display_language
-locale_get_display_name
-locale_get_display_region
-mysqli_get_client_stats
-mysqli_refresh
-mysqli_stmt_more_results
-mysqli_stmt_next_result
-net_get_interfaces
-normalizer_get_raw_decomposition
-openssl_cms_sign
-openssl_get_curve_names
-openssl_pkcs7_read
-openssl_pkey_derive
-openssl_x509_verify
-pcntl_unshare
-pg_lo_truncate
-phpdbg_get_executable
-pspell_config_data_dir
-pspell_config_dict_dir
-pspell_config_repl
-pspell_new_config
-realpath_cache_get
-sapi_windows_cp_conv
-sapi_windows_cp_get
-sapi_windows_set_ctrl_handler
-socket_addrinfo_bind
-socket_addrinfo_explain
-sodium_crypto_aead_aes256gcm_keygen
-timezone_location_get
```
`180` further moves out:
```
-curl_share_close
-curl_share_errno
-datefmt_set_timezone
-dns_check_record
-enchant_broker_dict_exists
-enchant_broker_free
-enchant_broker_free_dict
-enchant_broker_get_error
-enchant_broker_set_dict_path
-enchant_broker_set_ordering
-enchant_dict_add_to_session
-enchant_dict_check
-enchant_dict_get_error
-enchant_dict_store_replacement
-fdatasync
-get_mangled_object_vars
-gmp_popcount
-gmp_rootrem
-hash_hmac_algos
-imageaffinematrixconcat
-imageaffinematrixget
-imagecolorclosesthwb
-imagecropauto
-imageopenpolygon
-imagesetinterpolation
-imap_fetchmime
-imap_gc
-imap_getacl
-imap_getsubscribed
-imap_lsub
-imap_rfc822_write_address
-imap_set_quota
-imap_setacl
-imap_undelete
-inflate_add
-inflate_get_status
-inflate_init
-json_validate
-ldap_exop
-ldap_parse_reference
-locale_get_default
-locale_get_primary_language
-locale_get_region
-msg_queue_exists
-msg_set_queue
-msgfmt_format
-mysqli_get_connection_stats
-mysqli_get_links_stats
-mysqli_get_warnings
-mysqli_savepoint
-mysqli_stmt_attr_set
-mysqli_stmt_error_list
-mysqli_savepoint
-mysqli_stmt_attr_set
-mysqli_stmt_error_list
-normalizer_is_normalized
-openssl_get_cert_locations
-openssl_pbkdf2
-password_algos
-pcntl_setpriority
-pg_escape_identifier
-phpdbg_start_oplog
-posix_access
-posix_mknod
-posix_setrlimit
-pspell_add_to_personal
-pspell_config_create
-pspell_config_personal
-pspell_save_wordlist
-readline_callback_handler_install
-readline_callback_handler_remove
-readline_callback_read_char
-snmp2_getnext
-snmp2_walk
-snmp3_get
-snmp3_real_walk
-snmp_read_mib
-snmp_set_oid_output_format
-snmpgetnext
-socket_addrinfo_connect
-sodium_base642bin
-sodium_bin2base64
-sodium_crypto_aead_aes256gcm_decrypt
-sodium_crypto_aead_aes256gcm_encrypt
-sodium_crypto_aead_chacha20poly1305_encrypt
-sodium_crypto_aead_chacha20poly1305_ietf_decrypt
-sodium_crypto_aead_chacha20poly1305_ietf_encrypt
-sodium_crypto_aead_xchacha20poly1305_ietf_decrypt
-sodium_crypto_aead_xchacha20poly1305_ietf_encrypt
-sodium_crypto_box_keypair_from_secretkey_and_publickey
-sodium_crypto_box_open
-sodium_crypto_box_seal_open
-sodium_crypto_box_secretkey
-sodium_crypto_pwhash_str_verify
-sodium_crypto_sign_publickey_from_secretkey
-sodium_crypto_sign_seed_keypair
-sodium_crypto_stream
-stream_supports_lock
-timezone_version_get
-variant_cat
-variant_int
-variant_mul
-variant_sub
-zend_leak_variable
-zend_test_func
```

dune73 commented:
Thank you very much. Very good progress here.

A few remarks.

## frequency stats file

I would prefer not to commit `frequencylist.txt`, but 5h is a very long time. What options do you see?

## 933160

The second item in your bullet list is not clear to me. I think that anything marked as English word in 933161 should be added manually to 933160, but this is probably what you had in mind.

## 933161

Agreed

## 933150 and 933151

The limit of 100 is completely arbitrary. Adjusting it seems feasible to me. I'd aim for a similar size as we have now for 933150.

Am I getting you right, that your description brings the PHP function list that would be removed from 933150 if we raise the frequency limit from 100 to 150 and then further down to 180?

dune73 commented:
Do you have an update here for the meeting tonight? We should really get this done now.

M4tteoP commented:
Latest commit:
1. Adds `php-high-risk-functions.txt`. A manual list of high-risk functions becomes the second source to `933150`. 
2. `php-high-risk-functions.txt` is made of lots of `933160` entries, that are now moved from `933160` to `933150`. They are non-English words that we still want to detect at PL1.
3. Removes leftovers about `933150` in the script. It is a manual list.
4. Adds code to iterate over `933160` entries and add them to the stricter sibling `933161` if not already present.
5. Adds code to iterate over `php-high-risk-functions.txt` and adds them to `933150`.

Next steps / elements to be considered:
- We might want to create another file alongside `php-high-risk-functions.txt` and let the script generate `933150.ra` from it. The two files with manual lists would be closer, and these two `.ra` files would become fully generated.
- Double check `933160.ra` and `php-high-risk-functions.txt` entries
-  The spell checker based on WordNet is already used by this PR. With `extended spell.sh` I meant the [spell.sh script with also the hardcoded list of terms](https://owasp.slack.com/archives/CBKGH8A5P/p1688410762315539) added to the output of spell.sh suggested by @theseion. The examples were `basename`, `unset`, `syslog`, `symlink`. Now, not being matched by spell.sh:
   - `basename`: moved from `933161` to `933150` (high frequency) (PL3 -> PL1).
   - `unset`: removed. It is not found anymore as a PHP function 🧐.
   - `syslog`: moved from `933161` to `933151` (PL3 -> PL2).
   - `symlink`: moved from `933161` to `933151` (PL3 -> PL2).
- remove / gitignore `frequencylist.txt`
- Update rules description

dune73 commented on regex-assembly/933161.ra:
```diff
@@ -40,170 +21,58 @@ date
+unpack
 virtual
-wordwrap
+assert_options
```
I'd suggest to keep alphabetic order

dune73 commented:
I think there is going to be a consolidation project after CRS v4 comes out where we try to integrate all these automated word lists into the CRS toolchain. All work into developing an advanced scheme to minimize work in the future will have to be re-evaluated and streamlined. Likely being thrown out in favor of a new approach.

Personally, I would rather leave it dumb and describe the individual files, their role and their location.

But you do the work, so you get to decide.

theseion commented:
> The script takes about 5 hours to retrieve all the frequencies. I don't think we want to merge the frequencylist.txt, still committed it if anyone wants to play with it (Run command example ./php-dictionary-creator.sh --frequencylist frequencylist.txt --verbose).

I agree that we don't need it. However, I'd like for the lists and assembly files to include the script parameters in a header comment, so that when they are regenerated, it is clear which parameters were used.

> Updating 933161.ra requires running the toolchain. The script currently only prints a reminder about it. If we feel that it is in the scope of this script, this call could be also automated.

I think you should just try to run the toolchain and then report success or failure.


I don't fully understand what you are doing to the rules. From the comments in the rules file I gather that:

- 933150 includes symbols often used in attacks
- 933160 includes frequently used symbols
- 933151 includes low frequency symbols
- 933161 _should_ include symbols excluded from the other 3 rules due to FP risk

From that I gather that the symbols in all rules should be mutually exclusive (in the script you include symbols from 933160 in 933161). If this is not the case, then we need to update the rule comment (which needs an update anyway, IMO).

I tried to follow your discussion but it already took me quite some time to just read what you wrote. I trust that you've thought this through :)

theseion commented on util/php-dictionary-gen/php-dictionary-creator.sh:
```diff
@@ -0,0 +1,485 @@
+TMP_PHP_FUNCTIONS_RARE=$(mktemp)
+TMPDIR=$(mktemp -d)
+
+trap 'rm -rf $TMP_PHP_FUNCTIONS $TMP_PHP_FUNCTIONS_FREQUENCIES  $TMP_PHP_FUNCTIONS_FREQUENT $TMP_PHP_FUNCTIONS_RARE $TMP_ENGLISH_WORDS $TMPDIR' INT TERM EXIT
```
I wouldn't do this. You already have a date check in place. If the script fails at some point (after 2 hours because of a network error, for example) you don't want to rerun the full script but pick up where you left off. You should be able to pick up the work by skipping the entries that are young enough.

You could still add a `--clean` switch if you want.

M4tteoP replied:
All these files would still be not used in a second run, being random temp files generated.
We can start again from where the script failed (or from the last successful run) by setting an explicit path where to store the frequency list.
For context, I'm running the script with the following line: `./php-dictionary-creator.sh --frequencylist ~/frequencylist.txt`. Doing so, entries are individually stored into `frequencylist.txt` and kept for the next runs.

To automate it I could make `TMP_PHP_FUNCTIONS_FREQUENCIES` a fixed path, but then I would at least keep it in under the `util/` folder, and add it to `.gitignore`. No strong feelings about it, but I think that it makes sense to just be aware of where you store the frequency list file

theseion replied:
You could simply make `-f` a mandatory option / positional parameter. If there's no file, it will be created, otherwise it will be used like now. I think that would make it more clear how to use the script and will prevent the issue.

M4tteoP replied:
I did not report back here. Sorry.
By default now `util/php-dictionary-gen/frequencylist.txt` is used if no other frequency list file is specified via `--frequencylist`. Being still meant to be only a local file it has also been added to `.gitignore`

M4tteoP commented on regex-assembly/933161.ra:
```diff
@@ -2,208 +2,77 @@
 
 ##!+ i
-##!^ \b
+##!^ 
```
Thanks! `echo` is in charge of printing the prefix and `\b` was not escaped

M4tteoP commented:
> I don't fully understand what you are doing to the rules. 

Here is a full recap: 
- **933160** (PL2): List of high-risk English (and english-like) words
	- source: manual entries added to 933160.ra
	- note: Most of the 933160 entries are now moved to 933150 (via php-high-risk-functions.txt) being words now compatible with an English context.
- **933161** (PL3): List of PHP function names detected as English words by `spell.sh`
	- source: PHP source code, automated by the script (`php-dictionary-creator.sh`)
	- source2: 933161 also contains 933160 entries
- **933150** (PL1): List of PHP functions NOT detected as English words with HIGH frequency + PHP high-risk functions
	- source: PHP source code, automated by the script
	- source2: manual list of high risk functions (php-high-risk-functions.txt) all extracted from the old 933160 entries.
- **933151 (PL2)**: List of PHP functions NOT detected as English words with LOW frequency
	- source: PHP source code, automated by the script 
	- note: mutually exclusive with 933150

dune73 commented:
Thanks for your concerns @theseion. You are quite right, this is very complicated and the approach changed during the discussions. The problem we discovered was that the existing rules were not doing what they were describing to do. So this is not only about automation, but also overhauling existing rules, hence the move of high-risk function names from 933160 to 933150 (double checked, nothing was lost). The new distinction with regex 93316x for English words and 93315x for non-English words seems useful to me. 933150 and 933151 are covering all existing PHP function names and they are distinguished by the manually assessed risk as well as the new idea introduced by @jptosso to use frequency of use as a guidance.

What helps us is the good coverage of tests for 933150 and 933160. So any grave omission should be caught.

@M4tteoP : Thank you for your good writeup of the rule's functionality above. This is a good base for the new description of the rules. I can write the full description on this base if you want.
Also: Do you need anything else to finish the PR?

Do you have a solution for the problem with PHP function `dir` within English word `directory` mentioned above?

M4tteoP commented:
We still had two problems:
1. Lack of extended spell.sh that was leading to entries moving from 933161 to 933150 or 933151 (that means also moving from PL3 to PL1/PL2
2. problem with PHP function `dir` within English word `directory`. They were part of 933161 but now moved to 933150 (PL1):

To address both of them at once, I went for implementing the extended spell.sh, I hope that the implementation is kind o what you had in mind (cc @theseion). Adding to this list partial english words and common words like `basename` permits to not evaluate their frequency (typically very high) and end up in `933150`, but they rather take the other path of the mermaid and they end up in `933161`, were they used to be.

I squashed commits and rebased to make it easier to review and to address a conflict with the just merged https://github.com/coreruleset/coreruleset/pull/2302 (affecting `933161` regex). Please take a look at:

- [eec943d](https://github.com/coreruleset/coreruleset/pull/3273/commits/eec943d933eb360c4cdb7b4f86c00426432607c5) if you want to see/review the spell.sh extended implementation. Any suggestion about elements to be added is very welcomed.

Dealing with tests, I found  several duplicated tests for `933150`. Please take a look at:
- [328b1e6](https://github.com/coreruleset/coreruleset/pull/3273/commits/328b1e6bf029a48a4d1bc44ae5d9577e84323087) to see the fix.


I think it is mostly ready, trying to deal with the other failing tests now

M4tteoP commented:
There is one left failing test: `933161-4`. It is one that I fixed because the payload was a duplicate, but the description provided what I think the test was meant to match (`dl/*foo*/()`).
it seems that `dl` function in recent PHP versions is not allowed in web environments ([ref1](http://www.hackingwithphp.com/18/1/17/dont-use-dl), [ref2](https://www.php.net/manual/en/function.dl.php)), but I found the occurrence in the PHP repo not inside ZEND_FUNCTION(*), but PHP_FUNCTION(*).
I'm mainly wondering if we should look for also PHP_FUNCTION(*).
I'm running out of time for this week, but I will give it a go and see how many missing functions will be found.
If you have more knowledge about the PHP repo, please let me know.

If we are in a hurry to merge this PR, I would not consider it a blocker.

theMiddleBlue commented:
Wow, you've really put in a lot of effort here! Great job!

I'm just a bit worried about the regex causing a bunch of false positives. It's kind of a bummer since it even catches non-existent PHP function names. 

Like, for example, `preventivo123` (which means "estimate/quotation" in Italian). There's no injection stuff going on, just a simple `[a-z0-9]+` string, but it still ends up matching 933150 at **PL1** - PHP Injection Attack: High-Risk PHP Function Name Found.

```shell
$ curl -s -A 'test' 'http://sandbox.coreruleset.org/?a=preventivo123' \
   -H 'x-backend: nginx' \
   -H 'x-format-output: txt-matched-rules' \
   -H 'x-crs-paranoia-level: 1' \
   -H 'x-crs-version: pullrequest-3273'

933150 PL1 PHP Injection Attack: High-Risk PHP Function Name Found
949110 PL1 Inbound Anomaly Score Exceeded (Total Score: 5)
```

Allowing these sorts of false positives to linger at PL1 would turn into an absolute nightmare, and you know how I always mention that we need to be more stringent with injection context syntax. Remember, there's no way we can safeguard against something like `eval($_GET[foo])`, especially when we're talking about PL1.

I believe we could include a solution to tackle the most notable false positives by appending word boundaries `\b` after the PHP function name. What do you think about it?

IMO, this would be just a workaround. I think we should match an injection syntax before matching PHP function, at least at lower PLs.

dune73 commented:
Thank you. This sounds useful.

For the record: Our existing regex is good enough, because the function list is limited, but with the extension of the function list, the lacking injection syntax is blowing up in our face?

+1 on `\b`. More if needed.

theMiddleBlue commented:
> with the extension of the function list, the lacking injection syntax is blowing up in our face?

It is. Unfortunately, blocking PHP functions without an injection context is tricky. If we were to block a `[a-z0-9]+` string value at PL1, we would be overwhelmed by false positives.

theMiddleBlue commented:
I'd suggest giving it a shot with `\b` to avoid matching any other strings that begin with a familiar PHP function name, rather than adding exceptions.

theMiddleBlue commented:
(in the current nightly version, this false positive isn't popping up - IDK if it should and it doesn't work or because we changed something)

```shell
curl -s -A 'test' 'http://sandbox.coreruleset.org/?a=preventivo123' \
   -H 'x-backend: nginx' \
   -H 'x-format-output: txt-matched-rules' \
   -H 'x-crs-paranoia-level: 1'
```

dune73 commented:
OK. Back at this PR.

The `\b` idea does not work since the rule 933150 uses `@pm`. I think the only thing we can do is to remove function names that can be snippets of English words from 933150 and moving them to the manual high-risk list for 933160. We could also add them to 933161 but that's only PL3. Better using 933160 and explaining it in the file.

This includes the following terms:
* exp
* ord
* prev
* stat
* system (why the heck is this even here. This is an English word and should be in 933160 to begin with)

@M4tteoP can you implement this? Any help needed?

M4tteoP commented:
Agreed, I will do it by the weekend

dune73 commented:
That's very good news. If we can merge this on Monday before the meeting it would be great.

@theMiddleBlue Do you have anything else to add here? More concerns?

theMiddleBlue commented:
uhm... I suggest replacing the `@pm` operator with a regex (which shouldn't be too challenging with the crstoolchain). Currently, we're essentially shifting the issue to a higher level of paranoia, but I believe it's time to change our approach rather than just accepting false positives.

How difficult would it be to convert this into a regex?

In any case, can we postpone merging this until we've manually tested all of these rules?

thx

dune73 commented:
I do not think we are accepting FPs here. We are doing all we can within this architecture to avoid them. The basic strategy is: PHP functions names that do not exist in the English language are checked via PM and PHP function names that do exist in the English language either as word or as component of an English word are checked via regex.

What is the advantage of regex over pm when we make sure that the pm only gets those keywords that do not appear in the English language (and as we found out thanks to you: as snippets within words in the English language). I think this really addresses the core of the FP problem with the PM approach.

pm has the advantage that it's easier to read, easier to write and allegedly faster as well. If we stuff everything into a regex (like with the unix commands), we end up with a huge rule. Doable, but I would prefer to avoid it.

Finally, this is one of the roadblocks for CRS v4. Development is dragging along and we can't seem to finish things. I'd rather merge an optimized version of this than change the architecture before the goal and start over. If we optimize it and it still does not work, then I will agree with you.

dune73 commented:
For the record, I went through the php-function-names-933150.data in the PR and asked chatgpt for English words containing certain snippets where I was not sure. I added `unlink` and `substr` to he list above because of `unlinked` and `substring`. They seem to be used outside of IT contexts too.

theMiddleBlue commented:
I can't agree @dune73, `pm` is not the right operator for doing what we're trying to do here.
`pm` operator, is inadvertently matching English words containing PHP function names.

For example: `exp` PHP function is included in a tons of English words like

experience
expertise
exploration
exponential
expenditure
etc...

```shell
curl -v 'http://sandbox.coreruleset.org/?a=explosion' \
   -H 'x-backend: nginx' \
   -H 'x-format-output: txt-matched-rules' \
   -H 'x-crs-paranoia-level: 1' \
   -H 'x-crs-version: pullrequest-3273'

933150 PL1 PHP Injection Attack: High-Risk PHP Function Name Found
949110 PL1 Inbound Anomaly Score Exceeded (Total Score: 5)
```

I can find numerous examples similar to this one: `prev` or `ord` are other examples

Here we're blocking `?a=preview` at PL1!
```shell
$ curl -s 'http://sandbox.coreruleset.org/?a=preview' -H 'x-backend: nginx' -H 'x-format-output: txt-matched-rules' -H 'x-crs-paranoia-level: 1' -H 'x-crs-version: pullrequest-3273'

933150 PL1 PHP Injection Attack: High-Risk PHP Function Name Found
949110 PL1 Inbound Anomaly Score Exceeded (Total Score: 5)
```

**Furthermore, we're unintentionally blocking non-PHP functions, which is not our intended outcome.**

dune73 commented:
Instead of continuing the conversation ad nauseam, @theMiddleBlue and I hopped on a call.

We agreed on the course of action laid down in https://github.com/coreruleset/coreruleset/pull/3273#issuecomment-1701850262.

This should address known or expected FPs. If more FPs pop up with 933150, we should shift those keywords over to 933160 via the extended list. This includes stuff before we merge this PR and stuff that pops up during RC or after the release.

Redesigning the four rule of this PR, maybe consolidating into 1-2 regex rules could be an option. But @theMiddleBlue and I agreed this would throw us back at least another month. And we would be trusting our pattern identifying a php function call via the use of parenthesis etc. And we never know if there is a bypass in this regard. The pm operator in 933150 has its merits even if it is more prone to FP.

@M4tteoP promised to deliver an updated PR during the weekend.  @theMiddleBlue and I will review immediately afterwards, namely @theMiddleBlue will try out additional payloads to see if he can provoke more FPs. If things are back to green, I hope to be able to merge before the Monday meeting.

M4tteoP commented:
Done, the following terms have been moved out from `933150` and they are now part of `933160` and `933161` (being a stricter sibling).
- exp
- ord
- prev
- stat
- substr
- system
- unlink

Considering that `system` is back to `933160`, also the related tests have been moved back. The [conversion list of tests](https://github.com/coreruleset/coreruleset/pull/3273#issuecomment-1675490075) has been updated accordingly.

PTAL 🙏

theMiddleBlue commented on rules/php-function-names-933150.data:
```diff
@@ -1,44 +1,249 @@
+htmlentities
+htmlspecialchars
+htmlspecialchars_decode
+idate
```
**idate**
FP like `?a=validate`
this should moved out

theMiddleBlue commented on rules/php-function-names-933150.data:
```diff
@@ -1,44 +1,249 @@
+mb_eregi
+mb_eregi_replace
+mb_parse_str
+md5
```
**md5**
I think this is dangerous for API that could contains `hash=md5` could lead to FPs
Maybe it's a good idea to move this out, isn't it?

theMiddleBlue commented on rules/php-function-names-933150.data:
```diff
@@ -1,44 +1,249 @@
+php_uname
+phpinfo
+phpversion
+popen
```
**popen**
same for md5, maybe it's a good idea to move this out

dune73 commented:
For the record: There is a probability > 0 that random payloads (-> base64!) will hit 933150. The longer the keywords the less likely this is. Yet we have several 5 letter keywords in the list (`mkdir`, `fopen`, `fputs`, `rtrim`) as well as the 4 letter `is_a`. While `_` is not in the character set for base64, the pure letter words are. We are currently OK with 5 letter words, though. If we start to see random hits on 933150 because of base64 or some other random input, we might have to shift all those keywords out or we we really re-design this entire rule group.
